src/slot.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AmpContext,
+  withAmpContext,
+} from './amp-context.js';
+import {useMountEffect} from './amp-react-utils.js';
+
+const {
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+} = preactHooks;
+
+
+/**
+ * @param {!Element} element
+ * @param {string} name
+ * @param {!Object|undefined} props
+ * @return {!ReactElement}
+ */
+export function createSlot(element, name, props) {
+  element.setAttribute('slot', name);
+  const slotProps = Object.assign({name}, props || {});
+  return preact.createElement(Slot, slotProps);
+}
+
+/**
+ * Slot component.
+ */
+export function Slot(props) {
+  const context = useContext(AmpContext);
+  const ref = useRef();
+  const slotProps = Object.assign({}, props, {ref});
+  useEffect(() => {
+    const slot = ref.current;
+    const assignedElements = getAssignedElements(props, slot);
+    slot.__assignedElements = assignedElements;
+
+    // Retarget slots and content.
+    if (props.retarget) {
+      // TBD: retargetting here is for:
+      // 1. `disabled` doesn't apply inside subtrees. This makes it more like
+      //    `hidden`. Similarly do other attributes.
+      // 2. Re-propagate click events to slots since React stops propagation.
+      //    See https://github.com/facebook/react/issues/9242.
+      assignedElements.forEach(node => {
+        // Basic attributes:
+        const { attributes } = slot;
+        for (let i = 0, l = attributes.length; i < l; i++) {
+          const { name, value } = attributes[i];
+          if (name == 'name') {
+            // This is the slot's name.
+          } else if (!node.hasAttribute(name)) {
+            // TBD: this means that attributes can be rendered only once?
+            // TBD: what do we do with style and class?
+            node.setAttribute(name, value);
+          }
+        }
+        // Boolean attributes:
+        node.disabled = slot.hasAttribute('disabled');
+        node.hidden = slot.hasAttribute('hidden');
+        toggleAttribute(node, 'selected', slot.hasAttribute('selected'));
```

Comment:
What's this for?

Reply:
I think we will eventually switch to an explicit mapping for slot -> assignedNode attribute propagation. We currently do `disabled` because it doesn't inherit. We do `hidden` more as an signal for AMP for our hidden-observer. The `selected`, `expanded` and such are more similar to `disabled`. But I think we'll eventually want to make this mapping explicit like we do right now in AMP-binding's `opts: {attrs: {...}}` structure. Probably that mapping code could be even reused.

Reply:
Should this be a part of the AmpBaseElement options? It doesn't seem scalable to have sync every attribute between slot and children.

Reply:
Yes. I think this is where we are headed.